picoCTF-web/api/problem.py

```diff
@@ -307,7 +307,7 @@ def assign_instance_to_team(pid, tid=None, reassign=False):
     settings = api.config.get_settings()
     if settings["shell_servers"]["enable_sharding"]:
         available_instances = list(
-            filter(lambda i: i["server_number"] == team["server_number"],
+            filter(lambda i: i.get("server_number") == team.get("server_number", 1),
```

Comment:
Why does the instance `.get` not need a default? This code path simultaneously allow a team doc to not have server_number (so that is a scenario when someone turns on sharding afterwards) and yet fails mysteriously because the existing problem instances do not have server_number and so `None != 1`.

The only sane semantics should be "for any doc that could have server_number but doesn't, the default is 1".

Reply:
Actually, I just realized that even supporting the "the initially-off-and-then-on scenario" without database work is a bit harder because it is not safe to assume that any team doc that doesn't have `server_number` can be assumed to be `1`. For example, if the first step is smaller than the team number, then some team(s) will need to a default `server_number` of `2` or above.

This highlights the conceptual complexity (which is bad) of letting `server_number` ever be optional---even computing the default value is complex.

I currently think this is a good semantics to support this scenario (and again, this is the only scenario that we should support not):
1. Any problem instance that does not have `server_number` defaults to a `server_number` of `1`.
2. Any team that does not have `server_number` should use `assign_server_number()` to compute the default value.
3. The current code in the PR already ensures that all `shell_servers` docs have a `server_number`.

What do you think?

Reply:
See [above comment](#discussion_r219682110) about `.get` defaults

Once I figure out how/where to assign un-assigned teams, the line
` filter(lambda i: i.get("server_number") == team.get("server_number", 1),`
 will really be acting as an extra guard against assignment of instances of a server that does not have a `server_number` yet.